Move ct_lookup in bpf_host.c to a separate tailcall

gentoo-root opened the pull request:
Move ct_lookup in bpf_host.c to a separate tailcall to reduce verifier complexity.

Fixes: #17142

gentoo-root committed:
bpf: Initialize is_untracked_fragment just before usage

To prepare for splitting ipv4_host_policy_ingress, move the
initialization of the is_untracked_fragment variable closer to its only
usage.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Split ipv{4,6}_host_policy_ingress

Split the conntrack lookup part of the host policy ingress handlers into
a separate function to prepare for making it a separate tailcall. Use
struct ct_buffer{4,6} for communication between two parts of the old
function to prepare for passing the lookup result via a BPF map, similar
to how it's done in bpf_lxc.c.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Split ipv{4,6}_host_policy_egress

Split the conntrack lookup part of the host policy egress handlers into
a separate function to prepare for making it a separate tailcall. Use
struct ct_buffer{4,6} for communication between two parts of the old
function to prepare for passing the lookup result via a BPF map, similar
to how it's done in bpf_lxc.c.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Don't call send_drop_notify_error twice in handle_ipv4

When handle_ipv4 returns an error, send_drop_notify_error_ext is called
by tail_handle_ipv4, hence no need to call send_drop_notify_error from
handle_ipv4 itself.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Deduplicate ct_lookup4 in ipv4_host_policy_egress

ct_lookup4 is called by both ipv4_host_policy_egress_lookup and
whitelist_snated_egress_connections. Both lookups are the same and can
be replaced with ipv4_host_policy_egress_lookup. Remove the lookup part
from whitelist_snated_egress_connections and reuse the lookup that is
already present in ipv4_host_policy_egress.

While at it, also replace the error check after ct_create4 from IS_ERR
to ret < 0. ct_create4 only returns 0 or DROP_CT_CREATE_FAILED, and
having a mix of different error checks confuses the verifier when it
can't prove that CTX_ACT_DROP is not a possible outcome.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Split handle_ipv4 in bpf_host.c after ct_lookup4

In order to reduce verifier complexity, split handle_ipv4 into two
functions chained by a tailcall. The first part will stop after doing
ct_lookup4 of the host firewall, and the second part will pick up the
results from a BPF map and go on.

The tailcall is saved, and the map is not used if host firewall is
disabled, so the overhead is minimal in this case.

Also move the "verifier workaround" call to revalidate_data in
handle_ipv4 to the ENABLE_HOST_FIREWALL section, as its presence at the
previous location now makes clang miscompile code and generate 32-bit
zero extension on pkt and pkt_end pointers, which is not accepted by the
verifier. Removing the workaround altogether leads to another verifier
error: "R5 invalid mem access 'scalar'".

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Split handle_ipv6 in bpf_host.c after ct_lookup6

In order to reduce verifier complexity, split handle_ipv6 into two
functions chained by a tailcall. The first part will stop after doing
ct_lookup6 of the host firewall, and the second part will pick up the
results from a BPF map and go on.

The tailcall is saved, and the map is not used if host firewall is
disabled, so the overhead is minimal in this case.

Similar to the previous commit, also move the "verifier workaround"
call to revalidate_data in handle_ipv6 to the ENABLE_HOST_FIREWALL
section.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

pchaigno commented on bpf/lib/host_firewall.h:
```diff
@@ -256,22 +256,20 @@ whitelist_snated_egress_connections(struct __ctx_buff *ctx, __u32 ipcache_srcid,
-			ret = ct_create4(get_ct_map4(&tuple), &CT_MAP_ANY4,
-					 &tuple, ctx, CT_EGRESS, &ct_state_new,
+		ct_buffer->ret = ct_lookup4(get_ct_map4(tuple), tuple, ctx, l4_off,
+					    CT_EGRESS, &ct_buffer->ct_state, &ct_buffer->monitor);
```
I don't understand why I introduced a second CT lookup on in the same BPF program with the same tuple and same direction :thinking: Do you?

We should probably optimize that in a subsequent PR

gentoo-root replied:
Well, at least we don't do both, it's either one or the other, but indeed, it looks like they can be easily squashed together.

gentoo-root replied:
Did this cleanup in this PR, it allowed to get rid of some obscure flows, thanks for this great idea!

pchaigno commented on bpf/lib/host_firewall.h:
```diff
@@ -351,35 +367,28 @@ __ipv4_host_policy_egress(struct __ctx_buff *ctx, struct iphdr *ip4,
+		return ret;
 
 	if (src_id != HOST_ID) {
 #  ifndef ENABLE_MASQUERADE
```
I'm not sure I understand the rationale for commit `bpf: Simplify ipv4_host_policy_egress`. It results into more code and the `ENABLE_MASQUERADE` guard is now spread into two places.

We should at least add a comment before this line to explain why there's a `ENABLE_MASQUERADE` here. It may be enough to refer to the comment in `whitelist_snated_egress_connections`.

gentoo-root replied:
> the rationale

The goal is to allow ipv4_host_policy_egress_lookup to be called separately, not as part of ipv4_host_policy_egress, without putting a burden on the caller to check whitelist_snated_egress_connections. Imagine how the call to ipv4_host_policy_egress_lookup would look like if the call to whitelist_snated_egress_connections wasn't a part of it.

> ENABLE_MASQUERADE guard is now spread into two places.

It's temporary and patched up in a subsequent commit. I can probably avoid that if I introduce fill_trace right from the start.

gentoo-root replied:
That commit was dropped altogether and replaced with another one. Handling of whitelist_snated_egress_connections is much cleaner now.

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -608,22 +617,128 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 }
 
+static __always_inline int
+handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
```
Any reason to move this function to the bottom? If you keep it right after `resolve_srcid_ipv4`, then that's a lot less code movement and GitHub may even be able to recognize that huge portions of code didn't move. That would avoid many conflicts on backports and ease use of git blaming in the future.

gentoo-root replied:
> Any reason to move this function to the bottom?

Yes, the functions must be declared before they are called. The call graph used to look like:

```
tail_handle_ipv4 -> handle_ipv4 -> tail_handle_ipv4_cont -> handle_ipv4_cont
```

which meant that I either had to put the definition of handle_ipv4_cont above handle_ipv4, or at least add a forward declaration (which I thought would be uglier). However, after I fixed some bug, the call graph now looks like:

```
                 /-> handle_ipv4
tail_handle_ipv4
                 \-> tail_handle_ipv4_cont -> handle_ipv4_cont
```

which essentially means I can now put handle_ipv4 before handle_ipv4_cont and make the diff readable.

gentoo-root replied:
Function order restored.

pchaigno commented on bpf/lib/host_firewall.h:
```diff
@@ -372,23 +393,21 @@ ipv4_host_policy_egress(struct __ctx_buff *ctx, __u32 src_id,
 
-	if (src_id != HOST_ID) {
-#  ifndef ENABLE_MASQUERADE
+	if (fill_trace) {
```
Is this `fill_trace` variable actually needed? It seems to be always set when we reach this point as it's set when `ret >= 0`. If it is actually needed, it may be best to move its introduction to a separate commit to explain the rationale.

gentoo-root replied:
Not always set: if ENABLE_MASQUERADE is defined and src_id != HOST_ID, ipv4_host_policy_egress_lookup returns CTX_ACT_OK and doesn't set any trace fields. I'll try to move it to the "bpf: Simplify ipv4_host_policy_egress" commit.

gentoo-root replied:
fill_trace ditched from everywhere (thanks to the ct_lookup4 squash in whitelist_snated_egress_connections, that you suggested).

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -325,22 +316,125 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
 }
 
+static __always_inline int
+handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host)
```
Same as for its IPv4 counterpart.

pchaigno reviewed (commented):
Changes look good to me and the commit history is nice! :rocket: 

I have a couple comments below. Most notably, I'd like to reduce the code movements in the last two commits. That will be painful during future backports and git blaming sessions.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -426,71 +531,66 @@ resolve_srcid_ipv4(struct __ctx_buff *ctx, __u32 srcid_from_proxy,
-	struct remote_endpoint_info *info = NULL;
-	__u32 __maybe_unused remote_id = 0;
-	struct endpoint_info *ep;
+	__u32 from_host_raw = ctx_load_meta(ctx, CB_FROM_HOST);
```
One thing of note here is that in current code, `from_host` is a constant. So the compiler can effectively optimize ~half of the code out. Now we need to include support for both cases ... that doesn't feel like a win, complexity-wise?

gentoo-root replied:
That's a very valid point. I think I can fix it by using two tailcalls, just like CILIUM_CALL_IPV4_FROM_{HOST,NETDEV} already do, to keep this optimization for handle_ipv4_cont.

gentoo-root replied:
Addressed this one in the new version.

gentoo-root committed:
bpf: Let ipv{4,6}_host_policy_egress accept IP header

Currently, ipv{4,6}_host_policy_egress calls revalidate_data to get the
pointer to the IP header. However, if we analyze from which contexts
it's called, we realize that in most cases this pointer is already
available, and the call to revalidate_data can be optimized out:

1. handle_ipv{4,6} already have the pointer.

2. handle_to_netdev_ipv{4,6} also already have the pointer, moreover,
   handle_to_netdev_ipv4 calls revalidate_data and never uses its
   results - until this commit.

3. from_host_to_lxc doesn't have the pointer, but it always calls
   ipv{4,6}_host_policy_egress with src_id == HOST_ID, so this function
   will never return early before it would call revalidate_data. That
   means that the call to revalidate_data can be moved from
   ipv{4,6}_host_policy_egress to from_host_to_lxc without any
   performance penalty.

This commit essentially drops a few revalidate_data calls, no strings
attached.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -654,21 +841,59 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
+					   CILIUM_CALL_IPV4_CONT_FROM_NETDEV,
+					   tail_handle_ipv4_cont_from_netdev);
+	}
+	return ret;
```
If the invoke_tailcall_if() expands into a tail-call, we're missing the `send_drop_notify_error()` for the `DROP_MISSED_TAIL_CALL`.

I needed to fix up something similar recently in `handle_nat_fwd()`, maybe there's some inspiration there.

gentoo-root replied:
Ah, right, I missed that! At least the actual errors from cont are now handled properly :sweat_smile:

julianwiedmann commented on bpf/lib/host_firewall.h:
```diff
@@ -331,24 +417,50 @@ ipv4_host_policy_ingress(struct __ctx_buff *ctx, __u32 *src_id,
+	tuple = &tuple_on_stack;
+	memcpy(&ct_state_on_stack, &ct_buffer->ct_state, sizeof(ct_state_on_stack));
+	ct_state = &ct_state_on_stack;
+#endif
```
Note that this will "conflict" with https://github.com/cilium/cilium/pull/24504.

julianwiedmann replied:
-> that PR is now merged

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -243,6 +316,12 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
+	if (!from_host)
+		return CTX_ACT_OK;
+#endif /* !ENABLE_HOST_ROUTING */
+
```
Duplicated leftovers from conflict resolution? :thinking:

gentoo-root replied:
Seems so. I rebased and moved changes between patches a lot.

julianwiedmann reviewed (changes requested):
Looks great! Just a few small things below ...

There's also some typos in the commit description for `bpf: Deduplicate ct_lookup4 in ipv4_host_policy_egress` - it speaks of `ipv4_host_policy_egress_lookup()`, which doesn't exist at that point in the series.

gentoo-root reviewed (commented):
Thanks for reviewing it, I'll fix the remaining things.

julianwiedmann commented on bpf/lib/host_firewall.h:
```diff
@@ -204,14 +204,10 @@ ipv6_host_policy_ingress(struct __ctx_buff *ctx, __u32 *src_id,
-whitelist_snated_egress_connections(struct __ctx_buff *ctx, __u32 ipcache_srcid,
-				    struct trace_ctx *trace)
+whitelist_snated_egress_connections(struct __ctx_buff *ctx, struct ipv4_ct_tuple *tuple,
+				    int ct_ret)
```
nit - this could be `enum ct_status ct_ret`

gentoo-root replied:
ACK.

(For some reason this enum type is not popular in the code...)

julianwiedmann commented on bpf/lib/host_firewall.h:
```diff
@@ -223,39 +219,22 @@ whitelist_snated_egress_connections(struct __ctx_buff *ctx, __u32 ipcache_srcid,
+		int ret = ct_create4(get_ct_map4(tuple), &CT_MAP_ANY4,
+				     tuple, ctx, CT_EGRESS, &ct_state_new,
+				     false, false, false);
+		if (unlikely(ret < 0))
```
I believe this change isn't correct:
- in current code we extract a fresh CT tuple
- in new code, we re-use the previously extracted CT tuple. But that tuple has already undergone a `ct_lookup4()` with result `CT_NEW`, so its content has been flipped by `__ct_lookup4()`.

gentoo-root replied:
>in current code we extract a fresh CT tuple

No, it's also after ct_lookup4.

This change should be equivalent to the previous code (pure refactoring).

julianwiedmann replied:
Ah you're right of course, sorry for the noise.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -450,20 +454,25 @@ resolve_srcid_ipv4(struct __ctx_buff *ctx, __u32 srcid_from_proxy,
 	return src_id;
 }
 
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, __u32);
+	__type(value, struct ct_buffer4);
+	__uint(max_entries, 1);
+} CT_TAIL_CALL_BUFFER4 __section_maps_btf;
+
```
For my own education (I'll need to add a per-CPU map for some other use soon) - is there an obvious advantage to having separate IPv4/IPv6 maps? I would have instinctly gone with a single map, containing a union of the `ct_buffer4` / `ct_buffer6` structs.

gentoo-root replied:
I copied the same design we have in bpf_lxc.c. I don't see a real advantage over a union (type safety is equally absent when doing a lookup from a map).

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -652,21 +730,63 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
+	if (ret == CTX_ACT_OK) {
+		if (from_host)
+			invoke_tailcall_if(is_defined(ENABLE_HOST_FIREWALL),
+					   CILIUM_CALL_IPV4_CONT_FROM_HOST,
```
Probably just a matter of taste. But I would prefer to have the `ctx_store_meta(ctx, CB_SRC_LABEL, secctx)` here, and not hidden inside `handle_ipv4()`. It's already difficult enough to keep track of all the games we play with `CB_SRC_LABEL` ...

gentoo-root replied:
What about CB_FROM_HOST? Shall I move it out of handle_ipv4 too? If I do, it will be cumbersome to return 2 booleans in addition.

julianwiedmann replied:
`CB_FROM_HOST` should probably stay in it current place - as that's where we create its input data.

I'm just very sensitive to `CB_SRC_LABEL`, burned too many times already :smile:.

gentoo-root replied:
OK, I moved it and rebased.

(Please note that there is one more assignment to CB_SRC_LABEL in handle_ipv4 in the NAT46x64 shortcut that existed before my changes, but this should be fine, we do an unconditional tailcall right away.)

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -652,21 +730,63 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 	return ret;
 }
 
+declare_tailcall_if(is_defined(ENABLE_HOST_FIREWALL), CILIUM_CALL_IPV4_CONT_FROM_HOST)
+int tail_handle_ipv4_cont_from_host(struct __ctx_buff *ctx)
+{
+	return tail_handle_ipv4_cont(ctx, true);
+}
+
+declare_tailcall_if(is_defined(ENABLE_HOST_FIREWALL), CILIUM_CALL_IPV4_CONT_FROM_NETDEV)
+int tail_handle_ipv4_cont_from_netdev(struct __ctx_buff *ctx)
+{
+	return tail_handle_ipv4_cont(ctx, false);
+}
+
```
These are potentially executed as tail-calls, so we need to check for errors and call `send_drop_notify_error()`.

gentoo-root replied:
send_drop_notify_error_ext is in the common code (tail_handle_ipv4_cont), just like it was for tail_handle_ipv4 before my change.

julianwiedmann replied:
Missed that, thanks!

Note that the situation has slightly changed vs current code. In current code `tail_handle_ipv4_from_{host,netdev}()` are unconditional tail-calls, so we know that they *have* to handle drop/trace notifications.

With `declare_tailcall_if()` we also need to consider that the function gets inlined, and the surrounding code then expects to handle drop/trace notifications. Not a problem for drop notifications (as they are currently implemented through a tail-call, so control will be gone after the first drop notify). But trace notifications would get duplicated.

gentoo-root replied:
>Not a problem for drop notifications (as they are currently implemented through a tail-call, so control will be gone after the first drop notify)

I'd keep it as is then, at least for now. We also have a similar pattern in bpf_lxc.

I agree it's fragile, though. If the implementation for drop notifications is changed, it will become a problem that can easily slip unnoticed. This can be solved by passing handle_ipv4_cont (instead of tail_handle_ipv4_cont_from_{host,netdev}) to invoke_tailcall_if (if we forget about the arguments for a moment), but then the compiler will complain that a static function is never used (tail_handle_ipv4_cont_from_{host,netdev} when compiled without a tailcall). We need then to add __maybe_unused  to invoke_tailcall_if_0 (will it actually remove the function from the binary?).

julianwiedmann replied:
> I'd keep it as is then, at least for now. We also have a similar pattern in bpf_lxc.

oh yeah absolutely. I've been wanting to clean/fix some drop notification things anyway, maybe I'll tackle it along then at that point.

julianwiedmann reviewed (changes requested):
Sorry, took way too long to review this :disappointed:. Had a closer look, please check if my comments make sense ...

julianwiedmann reviewed (approved):
Couldn't spot anything else, and you nicely shot down all my earlier concerns :sweat_smile:.

Feel free to resolve my nits in a follow-on PR if it's easier. But as this needs a rebase anyway ...

pchaigno commented on bpf/lib/host_firewall.h:
```diff
@@ -284,15 +263,10 @@ ipv4_host_policy_egress(struct __ctx_buff *ctx, __u32 src_sec_identity,
-		/* Only enforce host policies for packets from host IPs. */
+	/* Only enforce host policies for packets from host IPs. */
+	if (src_sec_identity != HOST_ID &&
+	    (is_defined(ENABLE_MASQUERADE) || ipcache_srcid != HOST_ID))
```
The comment doesn't match the condition anymore. Maybe we need to move the comment in `whitelist_snated_egress_connections` here? We should probably also rename `whitelist_snated_egress_connections` to something like `track_snated_egress_connections`.

gentoo-root replied:
>Maybe we need to move the comment in whitelist_snated_egress_connections here?

I'll update this comment to make it more meaningful.

>We should probably also rename whitelist_snated_egress_connections to something like track_snated_egress_connections.

Why? I believe the meaning of the old name still holds. By creating a CT entry (the only thing this function does at the moment) we are whitelisting these connections from applying host policies to reply packets.

pchaigno replied:
I agree, but that intent is hard to understand if the code comment is moved to where the condition is.

pchaigno reviewed (approved):
I had a second, cursory look. One comment below, but I think it can be fixed up after the tests haven't passed (or after they are triaged) without needing to then rerun the full test suite (assuming you don't rebase in the process).

pchaigno commented on bpf/lib/host_firewall.h:
```diff
@@ -15,38 +15,50 @@
+			  struct ct_buffer6 *ct_buffer,
+			  struct trace_ctx *trace, __s8 *ext_err)
+{
+	struct ct_state ct_state_new = {};
```
I'm a bit confused by commit `bpf: Split ipv{4,6}_host_policy_egress` here. It says:

> Split the conntrack lookup part of the host policy egress handlers into a separate function to prepare for making it a separate tailcall.

but it seems the tail call was never introduced. Am I missing something @gentoo-root?

gentoo-root replied:
Maybe the phrasing was not the best... ipv6_host_policy_egress is split into ipv6_host_policy_egress_lookup and __ipv6_host_policy_egress. ipv6_host_policy_egress_lookup is called by the original caller of ipv6_host_policy_egress, and __ipv6_host_policy_egress continues after the tailcall. This is what's implemented in the next two commits. I don't think making ipv6_host_policy_egress_lookup an entirely separate (intermediate) tail call made sense back then, since it fit into handle_ipv6 by complexity. If it's not the case anymore, lookup is a separate function that can be made into another tailcall.
